cli/src/git_util.rs

```diff
@@ -137,18 +138,104 @@ fn get_ssh_keys(_username: &str) -> Vec<PathBuf> {
     paths
 }
 
-pub fn with_remote_git_callbacks<T>(ui: &Ui, f: impl FnOnce(git::RemoteCallbacks<'_>) -> T) -> T {
-    let mut callback = None;
+// Based on Git's implementation: https://github.com/git/git/blob/43072b4ca132437f21975ac6acc6b72dc22fd398/sideband.c#L178
+pub struct GitSidebandProgressMessageWriter {
+    display_prefix: &'static [u8],
+    suffix: &'static [u8],
+    scratch: Vec<u8>,
+}
+
+impl GitSidebandProgressMessageWriter {
+    pub fn new() -> Self {
```

Comment:
nit: `#[allow(clippy::new_without_default)]` to suppress the linter warning.

alternatively, we could pass `ui` to determine `.is_terminal()` by abusing `use_progress_indicator()`?